app/models/miq_ae_namespace.rb

```diff
@@ -8,25 +11,39 @@ class MiqAeNamespace < ApplicationRecord
                          /^commit_time/, /^commit_sha/, /^ref$/, /^ref_type$/,
                          /^last_import_on/, /^source/, /^top_level_namespace/].freeze
 
-  belongs_to :parent,        :class_name => "MiqAeNamespace",  :foreign_key => :parent_id
-  has_many   :ae_namespaces, :class_name => "MiqAeNamespace",  :foreign_key => :parent_id,    :dependent => :destroy
-  has_many   :ae_classes, -> { includes([:ae_methods, :ae_fields, :ae_instances]) },    :class_name => "MiqAeClass",      :foreign_key => :namespace_id, :dependent => :destroy
+  has_many :ae_classes, -> { includes([:ae_methods, :ae_fields, :ae_instances]) }, :class_name => "MiqAeClass", :foreign_key => :namespace_id, :dependent => :destroy
+
+  validates_format_of :name, :with    => /\A[\w\.\-\$]+\z/i,
+                             :message => N_("may contain only alphanumeric and _ . - $ characters")
+  validates :name, :presence => true, :uniqueness => {:scope => :ancestry, :case_sensitive => false}
+
+  alias ae_namespaces children
+  virtual_has_many :ae_namespaces
 
-  validates_presence_of   :name
-  validates_format_of     :name, :with    => /\A[\w\.\-\$]+\z/i,
-                                 :message => N_("may contain only alphanumeric and _ . - $ characters")
-  validates_uniqueness_of :name, :scope => :parent_id, :case_sensitive => false
+  after_save :clear_lookup_by_fqname_cache
 
   def self.lookup_by_fqname(fqname, include_classes = true)
     return nil if fqname.blank?
 
-    fqname   = fqname[0] == '/' ? fqname : "/#{fqname}"
-    fqname   = fqname.downcase
-    last     = fqname.split('/').last
+    @lookup_by_fqname_cache ||= {}
```

Comment:
I'd feel more comfortable if we didn't have a cache here.

Reply:
@kbrock Thanks for your review. Can you explain a bit more please?

Reply:
I think I explained it better in the other comment.

The cache is masking the actual problem: This is performing a query at each level.

The goal is to have a single query that will bring back the data you need.
It should bring back records across multiple domains if you need to do some cross domain logic (like ordering or something)
It should not bring back too many records that are not the final record of interest.

The find at each level of the tree is the problem.
Putting in a cache will hide this problem.
But you'll be paying for this underlying problem with (outdated?) models in memory and the initial queries that warmup that cache.

Reply:
moved to future PR